API/hermes_abi/hermes_abi.h

```diff
@@ -0,0 +1,270 @@
+#define HERMES_ABI_HERMES_ABI_H
+
+#include <assert.h>
+#include <stdalign.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#ifndef HERMES_EXPORT
+#ifdef _MSC_VER
+#define HERMES_EXPORT __declspec(dllexport)
+#else // _MSC_VER
+#define HERMES_EXPORT __attribute__((visibility("default")))
+#endif // _MSC_VER
+#endif // !defined(HERMES_EXPORT)
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/// Define the structure for references to pointer types in JS (e.g. string,
+/// object, BigInt).
+/// TODO: Replace jsi::PointerValue itself with this C implementation to
+/// eliminate pointer management overhead in the JSI wrapper.
+struct HermesABIManagedPointer;
+struct HermesABIManagedPointerVTable {
+  /// Pointer to the function that should be invoked when this reference is
+  /// released.
+  void (*invalidate)(HermesABIManagedPointer *);
+};
+struct HermesABIManagedPointer {
+  const HermesABIManagedPointerVTable *vtable;
+};
+
+enum HermesABIErrorCode {
+  HermesABIErrorCodeJSINativeException,
+  HermesABIErrorCodeJSError,
+};
+
+static_assert(
+    alignof(HermesABIManagedPointer) % 4 == 0,
+    "HermesABIManagedPointer must be at least aligned to pointer size.");
+
+/// Define simple wrappers for the different pointer types.
+#define HERMES_ABI_POINTER_TYPES(V) \
+  V(Object)                         \
+  V(Array)                          \
+  V(String)                         \
+  V(BigInt)                         \
+  V(Symbol)                         \
+  V(Function)                       \
+  V(ArrayBuffer)                    \
+  V(PropNameID)
+
+#define DECLARE_HERMES_ABI_POINTER_TYPE(name) \
+  struct HermesABI##name {                    \
+    HermesABIManagedPointer *pointer;         \
+  };                                          \
+  struct HermesABI##name##OrError {           \
+    uintptr_t ptrOrError;                     \
+  };
+
+HERMES_ABI_POINTER_TYPES(DECLARE_HERMES_ABI_POINTER_TYPE)
+#undef DECLARE_HERMES_ABI_POINTER_TYPE
+
+#define HERMES_ABI_TRIVIAL_OR_ERROR_TYPES(V) \
+  V(Bool, bool)                              \
+  V(Uint64, uint64_t)                        \
+  V(Uint8Ptr, uint8_t *)                     \
+  V(SizeT, size_t)
+
+#define DECLARE_HERMES_ABI_TRIVIAL_OR_ERROR_TYPES(name, type) \
+  struct HermesABI##name##OrError {                           \
+    bool is_error;                                            \
+    union {                                                   \
+      type val;                                               \
+      uint16_t error;                                         \
+    } data;                                                   \
+  };
+
+HERMES_ABI_TRIVIAL_OR_ERROR_TYPES(DECLARE_HERMES_ABI_TRIVIAL_OR_ERROR_TYPES)
+#undef DECLARE_HERMES_ABI_TRIVIAL_OR_ERROR_TYPES
+
+struct HermesABIVoidOrError {
+  bool is_error;
+  uint16_t error;
+};
+
+/// Always set the top bit for pointers so they can be easily checked.
+#define HERMES_ABI_POINTER_MASK (1 << (sizeof(int) * 8 - 1))
```

Comment:
I wonder why `sizeof(int)`. Since it is related to a pointer, it should be `sizeof(intptr_t)`.